File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
     mergeSortedList(list1, list2) {
        let dummy = new ListNode(0);
        let tail = dummy;

        while (list1 !== null && list2 !== null) {
            if (list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }

        if (list1 !== null) {
            tail.next = list1;
        }
        if (list2 !== null) {
            tail.next = list2;
        }

        return dummy.next;
    }

    /**
     * Merge k sorted linked lists
     */
    mergeKLists(lists) {
        if (lists.length === 0) return null;

        let merged = lists[0];

        for (let i = 1; i < lists.length; i++) {
            merged = this.mergeSortedList(merged, lists[i]);
        }

        return merged;
    }
    
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {

        int count= 0;
        ListNode temp=head;
        while(count <k)
        {
            if(temp==null)
            {
                return head;
            }
            temp=temp.next;
            count++;
        }

        ListNode prevNode= reverseKGroup(temp,k);
        
        ListNode curr=head;
        
        count=0;
        while(count<k)
        {
            ListNode next = curr.next;
            curr.next=prevNode;
            prevNode=curr;
            curr=next;
            count++;
        }

        return prevNode; 
    }
}
